scenario/scripts/snapshot.py

```diff
@@ -0,0 +1,752 @@
+        os.makedirs(os.path.dirname(filepath), exist_ok=True)
+        try:
+            fetch_file(
+                target,
+                container_name=container_name,
+                model=model,
+                remote_path=remote_path,
+                local_path=filepath,
+            )
+
+        except RuntimeError as e:
+            logger.error(e)
+
+    return mounts
+
+
+def get_container(
+    target: Target,
+    model,
+    container_name: str,
+    container_meta,
+    fetch_files: Optional[List[Path]] = None,
+    temp_dir_base_path: Path = SNAPSHOT_TEMPDIR_ROOT,
+) -> Container:
+    remote_client = RemotePebbleClient(container_name, target, model)
+    plan = remote_client.get_plan()
+
+    return Container(
+        name=container_name,
+        _base_plan=plan,
+        can_connect=remote_client.can_connect(),
+        mounts=get_mounts(
+            target,
+            model,
+            container_name,
+            container_meta,
+            fetch_files,
+            temp_dir_base_path=temp_dir_base_path,
+        ),
+    )
+
+
+def get_containers(
+    target: Target,
+    model,
+    metadata: Optional[Dict],
+    fetch_files: Dict[str, List[Path]] = None,
+    temp_dir_base_path: Path = SNAPSHOT_TEMPDIR_ROOT,
+) -> List[Container]:
+    fetch_files = fetch_files or {}
+    logger.info("getting containers...")
+
+    if not metadata:
+        logger.warning("no metadata: unable to get containers")
+        return []
+
+    containers = []
+    for container_name, container_meta in metadata.get("containers", {}).items():
+        container = get_container(
+            target,
+            model,
+            container_name,
+            container_meta,
+            fetch_files=fetch_files.get(container_name),
+            temp_dir_base_path=temp_dir_base_path,
+        )
+        containers.append(container)
+    return containers
+
+
+def get_status_and_endpoints(
+    target: Target, model: Optional[str]
+) -> Tuple[Status, Tuple[str, ...]]:
+    logger.info("getting status...")
+
+    status = _juju_run(f"status --relations {target}", model=model)
+    app = status["applications"][target.app_name]
+
+    app_status_raw = app["application-status"]
+    app_status = app_status_raw["current"], app_status_raw.get("message", "")
+
+    unit_status_raw = app["units"][target]["workload-status"]
+    unit_status = unit_status_raw["current"], unit_status_raw.get("message", "")
+
+    relations = tuple(app["relations"].keys())
+    app_version = app.get("version", "")
+    return Status(app=app_status, unit=unit_status, app_version=app_version), relations
+
+
+def _cast(value: str, _type):
+    if _type == "string":
+        return value
+    elif _type == "integer":
+        return int(value)
+    elif _type == "number":
+        return float(value)
+    elif _type == "boolean":
+        return value == "true"
+    elif _type == "attrs":  # TODO: WOT?
+        return value
+    else:
+        raise ValueError(_type)
```

Comment:
I see that this function is used only once throughout this entire file. Perhaps it would be better to roll it into the `get_config` function instead?